pkg/cluster/ClusterNoteService.go

```diff
@@ -0,0 +1,235 @@
+	clusterNoteService := &ClusterNoteServiceImpl{
+		clusterNoteRepository:        repository,
+		clusterNoteHistoryRepository: repositoryHistory,
+		userRepository:               userRepository,
+		logger:                       logger,
+	}
+	return clusterNoteService
+}
+
+func (impl *ClusterNoteServiceImpl) GenerateClusterDescription(bean *ClusterNoteBean, clusterBean *ClusterBean, isUpdate bool) (*ClusterDescriptionBean, error) {
+	clusterDescription := &ClusterDescriptionBean{}
+	if !isUpdate {
+		userDetails, err := impl.userRepository.GetById(int32(clusterBean.CreatedBy))
+		if err != nil {
+			impl.logger.Errorw("user service err, GetById", "err", err, "id", clusterBean.CreatedBy)
+			return nil, err
+		}
+		clusterDescription.ClusterCreatedOn = &clusterBean.CreatedOn
+		clusterDescription.ClusterCreatedBy = userDetails.EmailId
+		clusterDescription.ClusterId = clusterBean.Id
+		clusterDescription.ClusterName = clusterBean.ClusterName
+	}
+	if bean != nil {
+		userDetails, err := impl.userRepository.GetById(int32(bean.UpdatedBy))
+		if err != nil {
+			impl.logger.Errorw("user service err, GetById", "err", err, "id", bean.UpdatedBy)
+			return nil, err
+		}
+		clusterDescription.Id = bean.Id
+		clusterDescription.Description = bean.Description
+		clusterDescription.CreatedOn = &bean.UpdatedOn
+		clusterDescription.CreatedBy = userDetails.EmailId
+	}
+	return clusterDescription, nil
+}
+
+func (impl *ClusterNoteServiceImpl) Save(bean *ClusterNoteBean, userId int32) (*ClusterNoteBean, error) {
+	existingModel, err := impl.clusterNoteRepository.FindByClusterId(bean.ClusterId)
+	if err != nil && err != pg.ErrNoRows {
+		impl.logger.Error(err)
+		return nil, err
+	}
+	if existingModel.Id > 0 {
+		impl.logger.Errorw("error on fetching cluster, duplicate", "id", bean.ClusterId)
+		return nil, fmt.Errorf("cluster note already exists")
+	}
+
+	model := &repository.ClusterNote{
+		ClusterId:   bean.ClusterId,
+		Description: bean.Description,
+	}
+
+	model.CreatedBy = userId
+	model.UpdatedBy = userId
+	model.CreatedOn = time.Now()
+	model.UpdatedOn = time.Now()
+
+	err = impl.clusterNoteRepository.Save(model)
+	if err != nil {
+		impl.logger.Errorw("error in saving cluster note in db", "err", err)
+		err = &util.ApiError{
+			Code:            constants.ClusterCreateDBFailed,
+			InternalMessage: "cluster note creation failed in db",
+			UserMessage:     fmt.Sprintf("requested by %d", userId),
+		}
+	}
+	bean.Id = model.Id
+	bean.UpdatedBy = int(model.UpdatedBy)
+	bean.UpdatedOn = model.UpdatedOn
+	// audit the existing description to cluster audit history
+	clusterAudit := &repository.ClusterNoteHistory{
+		NoteId:      bean.Id,
+		Description: bean.Description,
+	}
+	clusterAudit.CreatedBy = userId
+	clusterAudit.UpdatedBy = userId
+	clusterAudit.CreatedOn = time.Now()
+	clusterAudit.UpdatedOn = time.Now()
+	err = impl.clusterNoteHistoryRepository.SaveHistory(clusterAudit)
+	if err != nil {
+		return nil, err
```

Comment:
Don't return err, only log it if there is error in saving history

Reply:
implemented 👍